### PR TITLE
H200: TTA mirror-aug control on non-mirror-trained H189

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Thorfinn's H192 (TTA mirror-aug on H112 checkpoint, currently running) tests whether test-time mirror augmentation improves test metrics on a **mirror-trained** model. H200 is the **mechanism control**: apply identical TTA to a **non-mirror-trained** checkpoint (askeladd's H189 AdamW+tau_y=3.0, run `bp12b1cc`).

- If H192 (mirror-trained + TTA) shows test_WSS gain AND H200 (non-mirror-trained + TTA) does NOT → TTA gain comes from **learned mirror invariance** (mechanism-specific)
- If both show similar gain → TTA is generic averaging benefit (applicable to any checkpoint)
- If both show gain but H192 > H200 → partial: both averaging + invariance contribute

This directly informs future strategy: if generic, TTA can be applied to any checkpoint; if mirror-specific, compounds involving mirror-aug are the only path.

## Instructions

**This is a pure eval sprint — NO training. No source file modifications.**

### Step 1: Locate H189 checkpoint
- W&B run: `bp12b1cc` (W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`)
- Pull the best val_abupt EMA checkpoint (EP9 or terminal, whichever had lower val_abupt from W&B)

### Step 2: TTA mirror-aug pipeline

For each test mesh:
1. **Pass A**: forward pass on ORIGINAL geometry
2. **Pass B**: forward pass on **left-right mirrored** geometry:
   - Mirror inputs: negate y, normal_y, tau_y in surface data; y in volume data
   - Invariant inputs (do NOT negate): cp, tau_x, tau_z, sdf, volume_pressure
3. **Un-mirror Pass B outputs**: negate the y-component of any output that has lateral sign sensitivity (WSS_y, velocity_y equivalent)
4. **Average**: pred_final = 0.5 × pred_A + 0.5 × pred_B

Run three eval modes:
- **Eval A** (original only, baseline)
- **Eval B** (mirrored only, ablation)
- **Eval C** (TTA average, the experiment)

```bash
python -m torch.distributed.run --nproc-per-node=8 \
  train.py \
  --eval_only \
  --checkpoint <path-to-H189-best-val-EMA> \
  --eval_mode original \
  --wandb_group h200-edward-tta-control \
  --run_name h200-h189-original

# repeat with --eval_mode mirrored and --eval_mode tta
```

If the train.py eval pipeline does not natively support `--eval_mode`, implement a thin Python wrapper that:
1. Loads the model weights
2. Runs both forward passes with the mirror flip applied manually (see `data/loader.py` for mirror logic — read it, do NOT modify it)
3. Averages outputs with correct sign-flip on y-axis components
4. Feeds averaged outputs to the standard metric computation

### Step 3: Report

For each eval mode (A, B, C):
- test_WSS, test_VP, test_SP per channel
- Delta C vs A (TTA gain)

Compare H200's delta C vs H192's delta (when H192 result available) for mechanism attribution.

## Baseline

H189 terminal (non-mirror-trained): test_WSS=7.120%, test_VP=3.664%, test_SP=3.949% — this is the starting point for H200 TTA, NOT the baseline gate.
H112 baseline (H192 TTA target): test_WSS=6.752%, test_VP=3.421%, test_SP=3.695%.

For H200 to produce a **merge-eligible result** standalone: test_WSS ≤ 6.752% — unlikely since H189 was 7.12%. The value is DIAGNOSTIC, not standalone merge.

Merge gate (for completeness):
- test_WSS ≤ 6.752%, test_VP ≤ 3.421%, test_SP ≤ 3.577%

Expected runtime: ~1h on 8 GPUs (3 evals × ~20min each).

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["jsdf4oww"],"primary_metric":{"name":"test_primary_tta/wall_shear_rel_l2_pct","value":8.3932},"test_metric":{"name":"test_primary_tta/wall_shear_rel_l2_pct","value":8.3932}}

## Results

H200 is the **mechanism control** for H192. We apply identical 2-pass mirror TTA (original + un-mirrored, averaged in normalized space) to the non-mirror-trained H189 checkpoint (`bp12b1cc`, EMA EP13). The original-mode eval reproduces the H189 baseline metrics to within 1e-5 (e.g. test_WSS 7.11961% vs published 7.11960%), confirming correct checkpoint loading and eval pipeline parity.

### Per-mode metrics

| Split | Mode | abupt | SP | WSS | VP | WSS_x | WSS_y | WSS_z |
|-------|------|------:|---:|----:|---:|------:|------:|------:|
| test  | original | 6.165 | 3.949 | 7.120 | 3.664 | 6.366 | 7.661 | 9.187 |
| test  | mirrored | 11.227 | 8.891 | 11.652 | 8.751 | 10.090 | 13.357 | 15.045 |
| test  | tta      | 7.733 | 5.649 | **8.393** | 5.486 | 7.405 | 9.281 | 10.844 |
| val   | original | 6.448 | 4.307 | 7.311 | 3.770 | 6.436 | 7.881 | 9.844 |
| val   | mirrored | 11.317 | 9.057 | 11.637 | 8.770 | 10.063 | 13.209 | 15.486 |
| val   | tta      | 7.863 | 5.791 | 8.477 | 5.397 | 7.407 | 9.325 | 11.396 |

### Delta (TTA − Original) on test

| Metric | Delta (pp) |
|--------|-----------:|
| abupt | +1.568 |
| WSS (vector) | +1.274 |
| WSS_x | +1.038 |
| WSS_y | +1.621 |
| WSS_z | +1.657 |
| SP | +1.700 |
| VP | +1.823 |

**All deltas are positive → TTA hurts every channel on this non-mirror-trained checkpoint.**

### Comparison to H189 baseline gate

| Metric | H189 baseline | H200 best (original) | H200 TTA | Merge gate |
|--------|-------------:|---------------------:|---------:|-----------:|
| test_WSS | 7.120 | 7.120 | 8.393 | ≤ 6.752 ❌ |
| test_VP  | 3.664 | 3.664 | 5.486 | ≤ 3.421 ❌ |
| test_SP  | 3.949 | 3.949 | 5.649 | ≤ 3.577 ❌ |

Merge gate not met (advisor flagged this expected: "DIAGNOSTIC, not standalone merge"). Value is mechanism attribution.

### Command

```bash
\
python -m torch.distributed.run --nproc-per-node=8 tta_mirror_eval.py \
  --checkpoint outputs/h189-bp12b1cc/checkpoint.pt \
  --config-yaml outputs/h189-bp12b1cc/config.yaml \
  --data-root /mnt/new-pvc/Processed/drivaerml_processed \
  --manifest data/split_manifest.json \
  --batch-size 4 \
  --eval-surface-points 65536 \
  --eval-volume-points 65536 \
  --amp-mode bf16 \
  --splits test,val \
  --eval-modes original,mirrored,tta \
  --wandb-group h200-edward-tta-control \
  --wandb-name edward/h200-h189-tta-mirror
```

Runtime: 133s (val) + 199s (test) = ~5.5 min on 8×H200 (well under the 1h budget). Two forward passes per batch (A=original, B=mirrored), shared across modes.

Peak memory: ~14 GB/GPU (DDP, bf16, batch=4, 65536 points).

W&B run: `jsdf4oww` (group `h200-edward-tta-control`)

### What happened

**H189 has zero learned mirror invariance.** The mirrored-only eval shows catastrophic degradation across every channel:
- SP: 3.95% → 8.89% (+4.94pp, 2.25× worse)
- WSS: 7.12% → 11.65% (+4.53pp)
- VP: 3.66% → 8.75% (+5.09pp, 2.39× worse)
- WSS_y (the most lateral-sensitive axis): 7.66% → 13.36% (+5.70pp)

Because the mirrored pass is so much worse than the original, averaging the two in normalized space is strictly worse than the original alone — the bad pass drags the average up. This is the unambiguous **negative** half of the mechanism control: a model that hasn't been trained with mirror augmentation has nothing for TTA to exploit.

**Mechanism attribution for H192 (pending):**
- **If H192 shows test_WSS gain** under TTA (e.g. ≤ ~6.7%) → the gain is **mechanism-specific**: it requires the model to have learned mirror invariance during training. TTA is not a generic averaging trick.
- **If H192 also degrades** → TTA mirror-aug is a dead end for our checkpoints, regardless of training augmentation.
- **If H192 degrades less than H200** → partial credit: mirror-training reduces the cost of the bad pass, but doesn't recover full averaging benefit.

The clean negative result on H200 makes H192's comparison highly informative either way — we now know the "control" floor: TTA on a mirror-naïve model costs ~1.3pp on WSS, so any H192 result better than +1.3pp shows the mirror training is doing real work.

### Pipeline notes

- `tta_mirror_eval.py` is a thin wrapper around the existing `EvalAccumulator` / `finalize_eval_accumulator` infra in `trainer_runtime.py`, so all metric semantics are bit-identical to `train.py`. No source files modified.
- Mirror operation flips: surface inputs y, n_y; volume inputs y. Mirror predictions un-flipped: surface tau_y (with the small mean-shift correction `-2*mu/sigma` since outputs are in normalized space). cp, tau_x, tau_z, sdf, volume pressure are invariant under y-reflection.
- TTA averaging is in **normalized** prediction space (before denormalization), which is what the loss is computed in. The 0.5 weighting is uniform.
- The original-mode eval reproduces H189's baseline to better than 1 ppm across all 7 primary metrics (e.g. test_WSS 7.119606 vs 7.119599), establishing pipeline parity.

### Suggested follow-ups

1. **Wait for H192 result.** The control's value is comparative; the H192 delta vs H200 is the load-bearing measurement. Sketched outcomes already covered in "Mechanism attribution" above.
2. **If H192 wins**: try TTA with stronger ensembling — N-pass with small rotational/translational perturbations on a mirror-trained checkpoint, since the invariance learning generalizes beyond the y-axis flip. Could compound on top of mirror TTA.
3. **If H192 loses by a similar margin to H200**: definitively close the TTA family. The asymmetry of DrivAerML (vehicle is left-right symmetric but flow conditions and small geometric details may not be exactly so) might mean true mirror invariance is unlearnable from current training data, regardless of augmentation.
4. **Diagnostic for H192**: when H192 lands, separately log the *mirrored-only* metric. If H192's mirrored eval is close to its original (say within 0.3pp), the model has actually learned the symmetry; if it's still 3-4pp off, mirror training was only partial and TTA averaging will be dominated by the bad pass just like H200.
5. **Not applicable here**: the H200 TTA result does not change H189's reported baseline — H189's published `test_primary/*` (the original-mode equivalent) stands.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1374 because ADVISOR: **H200 closed — CONTROL RESULT BANKED. Mechanism attribution NEGATIVE half clean.**

Excellent work — this is exactly the diagnostic value we wanted from a control. The pipeline parity reproduction (test_WSS 7.119606 vs published 7.119599, < 1ppm) establishes the eval infrastructure is correct, and the mirrored-only eval gives us a quantitative measurement of how much mirror invariance H189 learned: **catastrophically little** (+4.53pp test_WSS, +5.70pp WSS_y, +5.09pp VP).

### Finding banked: TTA mirror-aug requires explicit training augmentation

| Metric | H189 baseline (original) | H189 + mirror TTA | Delta |
|---|---:|---:|---:|
| test_WSS | 7.120 | 8.393 | **+1.273pp** (hurts) |
| test_VP | 3.664 | 5.486 | **+1.823pp** (hurts) |
| test_SP | 3.949 | 5.649 | **+1.700pp** (hurts) |

**Finding N (program-permanent)**: TTA mirror-aug on a non-mirror-trained model is **strictly worse** than the original-only eval. The mirrored pass produces predictions ~5pp worse on every channel, and uniform averaging drags the result toward the bad pass. This sets the CONTROL FLOOR for interpreting H192/H206/H209/H211: any TTA improvement on those mirror-trained models can only be attributed to mirror-aug training itself.

Your "Mechanism attribution" framework (in your last comment) is exactly the right way to think about the upcoming results:
- If H192 (TTA on H112, non-mirror-trained) shows delta similar to H200's −1.27pp → confirms TTA is generic averaging trick, mirror-aug training has zero residual value at inference
- If H206/H209/H211 (TTA on mirror-trained H183/H185/H190) show delta < +0.5pp → mirror-aug training reduces the "bad pass" gap, even if no net SOTA
- If any of H206/H209/H211 show delta NEGATIVE on test_WSS AND clear all 4 gates → SOTA candidate, mirror-aug TTA recovers basin

### Next assignment: H212

Mirror invariance profile across cohort EP13 checkpoints. Branch incoming.

Closing this PR. Result logged to research/EXPERIMENTS_LOG.md..
